Tree Widget & Property Grid: e2e tests on docker container

JonasDov opened the pull request:
Closes #614

grigasp commented on packages/itwin/property-grid/scripts/start-test-viewer-no-watch.js:
```diff
@@ -0,0 +1,24 @@
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+const crossEnv = require("cross-env");
```
maybe we could enhance existing `start-test-viewer.js` to accept a `--watch` argument?

JonasDov replied:
Changed

grigasp commented on packages/itwin/property-grid/scripts/run-e2e-with-docker.js:
```diff
@@ -0,0 +1,53 @@
+const dockerImageName = `${packageName}-e2e-test-image`
+const dockerContainerName = `${packageName}-e2e-test-container`
+
+const execute = command => new Promise((resolve, reject) => {
```
Our coding style suggests wrapping lambda args with parentheses.

```suggestion
const execute = (command) => new Promise((resolve, reject) => {
```

JonasDov replied:
Changed

grigasp commented on packages/itwin/property-grid/scripts/run-e2e-with-docker.js:
Replace all single quotes with double quotes

grigasp commented on packages/itwin/property-grid/scripts/run-e2e-with-docker.js:
```diff
@@ -0,0 +1,53 @@
+  const rootDirectory = path.resolve(currentDirectory, '../../../');
+  process.chdir(rootDirectory);
+} catch (err) {
+  console.error('Failed to change directory:', err);
```
I believe you wanted the following:

```suggestion
  console.error(`Failed to change directory: ${err}`);
```

JonasDov replied:
Changed

grigasp commented on packages/itwin/property-grid/scripts/run-e2e-with-docker.js:
```diff
@@ -0,0 +1,53 @@
+  .then(() => {
+    if (process.env.UPDATE_SNAPSHOTS) {
+      // Copy snapshots from docker container to the local repo
+      return execute(`docker cp ${dockerContainerName}:/workspaces/viewer-components-react/packages/itwin/${packageName}/src/e2e-tests/PropertyGrid.test.ts-snapshots ./packages/itwin/${packageName}/src/e2e-tests`);
```
Can we avoid using `PropertyGrid.test.ts-snapshots` folder name? It depends on test source file name...

JonasDov replied:
Removed

grigasp commented on packages/itwin/property-grid/scripts/run-e2e-with-docker.js:
This is a pretty generic script - maybe we could accept `packageName` as an argument and move the script to a common dir so other package can reuse it?

JonasDov replied:
Moved it to common folder

grigasp commented on packages/itwin/property-grid/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+# Downloads needed for git, npm and playwright to work
+RUN apt-get update \
+  && apt-get install -y curl \
+  && apt-get install -y libnss3 libatk-bridge2.0-0 libasound2 \
+  libcairo2 libcups2 libdbus-1-3 libatk1.0-0 libdrm2 libxrandr2 \
+  libglib2.0-0 libxkbcommon0 libnspr4 libatspi2.0-0 libgbm1\
+  libpango-1.0-0 libxcomposite1  libxdamage1  libxfixes3\
+  && curl -sL https://deb.nodesource.com/setup_18.x | bash - \
+  && apt-get install -y git nodejs locales zsh procps
```
do we really need all these deps?

grigasp replied:
The first part looks okay.

For the second part, `RUN npx playwright install-deps` definitely looks more future-proof than installing specific dependencies that may change with future releases of `playwright`. What was the failure?

grigasp replied:
Have you looked at using a docker image made specially for playwright - https://playwright.dev/docs/docker?

JonasDov replied:
Changed "FROM docker.io/library/debian:11" to "FROM mcr.microsoft.com/playwright:v1.36.2-jammy"

grigasp commented on packages/itwin/property-grid/Dockerfile:
This looks very generic. Maybe we could inject the package name using docker build args and reuse it across packages?

JonasDov replied:
Moved it to root folder

grigasp reviewed (changes requested):
Let's also try to remove `expect.toHaveScreenshot.maxDiffPixelRatio` config from `playwright.config.ts` - running the tests in a docker container should hopefully make it unnecessary.

JonasDov commented on apps/test-viewer/package.json:
```diff
@@ -74,14 +74,16 @@
     "rebuild": "npm run clean && npm run build",
-    "link:deps": "node ./scripts/linkWorkspaceDeps.js"
+    "link:deps": "node ./scripts/linkWorkspaceDeps.js",
+    "link:deps:no-watch": "node ./scripts/linkWorkspaceDeps.js --no-watch"
```
Changed

JonasDov commented on apps/test-viewer/scripts/linkWorkspaceDeps.js:
```diff
@@ -39,20 +40,22 @@ function linkPackages() {
-      }
-      lastChange = now;
-      setTimeout(() => {
+    if (watch){
```
Changed

JonasDov commented on packages/itwin/property-grid/playwright.config.ts:
```diff
@@ -26,7 +26,7 @@ export default defineConfig({
   workers: process.env.CI ? 1 : undefined,
   /* Reporter to use. See https://playwright.dev/docs/test-reporters */
-  reporter: process.env.CI ? [["html", { open: "never" }]] : "html",
+  reporter: [ ['html', { open: 'never'}]],
```
Changed

grigasp commented on packages/itwin/property-grid/scripts/start-test-viewer.js:
```diff
@@ -1,9 +1,18 @@
-const path = require('path').resolve(__dirname, '../../../../apps/test-viewer')
+const path = require("path").resolve(__dirname, "../../../../apps/test-viewer");
+const watch = process.argv[2] && process.argv[2] === "--watch";
+let options = {
+  shell: true,
+};
+let startDevPrefix = ":no-watch";
+if (watch) {
+  options = {};
+  startDevPrefix = "";
+}
```
I think it would be clearer if we just had this:
```
const scriptName = watch ? "start:dev" : "start:dev:no-watch";
const options = watch ? undefined : { shell: true };
```

JonasDov replied:
Changed

JonasDov replied:
Renamed

grigasp commented on Dockerfile:
I believe having a file called `Dockerfile` at the repo root is very unclear as the filename doesn't tell anything about what it's for. Maybe we could rename it to `e2e.Dockerfile` or something like that?

JonasDov replied:
Renamed

grigasp commented on Dockerfile:
```diff
@@ -0,0 +1,46 @@
+# Setup rush
+RUN npm install -g @microsoft/rush
+RUN rush install
+RUN rush build
```
Let's avoid building packages that we don't need 

```suggestion
RUN rush build -t test-viewer
```

JonasDov replied:
Changed

grigasp commented:
Just tried running the tests locally. A couple of observations:
- When running for the first time, it was taking quite a while. That's fine, but the console wasn't showing any progress and seemed like it was stuck. Could we forward container's log to host's console?
- When trying to run the tests the second time, I got the following output:
   ```
   > @itwin/property-grid-react@1.3.0 test:e2e
   > node ../../../common/scripts/run-e2e-with-docker.js property-grid
   
   
   #0 building with "default" instance using docker driver
   
   #1 [internal] load build definition from Dockerfile
   #1 transferring dockerfile: 1.67kB done
   #1 DONE 0.0s
   
   #2 [internal] load .dockerignore
   #2 transferring context: 76B done
   #2 DONE 0.0s
   
   #3 [internal] load metadata for docker.io/library/debian:11
   #3 DONE 3.7s
   
   #4 [ 1/11] FROM docker.io/library/debian:11@sha256:f33900927c0a8bcf3f0e2281fd0237f4780cc6bc59729bb3a10e75b0703c5ca7
   #4 DONE 0.0s
   
   #5 [internal] load build context
   #5 transferring context: 702.58kB 0.5s done
   #5 DONE 0.5s
   
   #6 [ 8/11] RUN rush install
   #6 CACHED
   
   #7 [ 2/11] RUN apt-get update   && apt-get install -y curl   && apt-get install -y libnss3 libatk-bridge2.0-0 libasound2   libcairo2 libcups2 libdbus-1-3 libatk1.0-0 libdrm2 libxrandr2   libglib2.0-0 libxkbcommon0 libnspr4 libatspi2.0-0 libgbm1  libpango-1.0-0 libxcomposite1  libxdamage1  libxfixes3  && curl -sL https://deb.nodesource.com/setup_18.x | bash -   && apt-get install -y git nodejs locales zsh procps
   #7 CACHED
   
   #8 [ 3/11] WORKDIR /workspaces/viewer-components-react/
   #8 CACHED
   
   #9 [ 4/11] COPY . .
   #9 CACHED
   
   #10 [ 6/11] RUN git config --local user.name imodeljs-admin
   #10 CACHED
   
   #11 [ 5/11] RUN git config --local user.email imodeljs-admin@users.noreply.github.com
   #11 CACHED
   
   #12 [ 7/11] RUN npm install -g @microsoft/rush
   #12 CACHED
   
   #13 [10/11] WORKDIR /workspaces/viewer-components-react/packages/itwin/property-grid
   #13 CACHED
   
   #14 [ 9/11] RUN rush build
   #14 CACHED
   
   #15 [11/11] RUN npx playwright install
   #15 CACHED
   
   #16 exporting to image
   #16 exporting layers done
   #16 writing image sha256:941f74ddb626929538c7482e18d1031c0f11dd1693a26e8365b70a7a9c31a888 done
   #16 naming to docker.io/library/property-grid-e2e-test-image done
   #16 DONE 0.0s
   
   What's Next?
     View a summary of image vulnerabilities and recommendations → docker scout quickview
   
   Error executing command: Error: Command failed: docker run --name property-grid-e2e-test-container -e UPDATE_SNAPSHOTS=1 property-grid-e2e-test-image
   docker: Error response from daemon: Conflict. The container name "/property-grid-e2e-test-container" is already in use by container "b3f3e716c36f6b0314f3c616c75f7a9a42b79f9960c1c0eef9e67200f6dde769". You have to remove (or rename) that container to be able to reuse that name.
   See 'docker run --help'.
   
   
   > @itwin/property-grid-react@1.3.0 test:e2e:local
   > playwright test
   
   
   
   
   Running 4 tests using 4 workers
     4 passed (44.8s)                                                                                                                                                                                                                                                                                                                               
   
   To open last HTML report run:
   
     npx playwright show-report
   
   [WebServer] (node:127) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.                                                                                                                                                           
   [WebServer] (node:127) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.            
   ```

   It looks like it succeeded but what's up with that error? 
   ```
   Error executing command: Error: Command failed: docker run --name property-grid-e2e-test-container -e UPDATE_SNAPSHOTS=1 property-grid-e2e-test-image
   docker: Error response from daemon: Conflict. The container name "/property-grid-e2e-test-container" is already in use by container "b3f3e716c36f6b0314f3c616c75f7a9a42b79f9960c1c0eef9e67200f6dde769". You have to remove (or rename) that container to be able to reuse that name.
   See 'docker run --help'.
   ```

JonasDov commented:
> Just tried running the tests locally. A couple of observations:
> 
> * When running for the first time, it was taking quite a while. That's fine, but the console wasn't showing any progress and seemed like it was stuck. Could we forward container's log to host's console?
> * When trying to run the tests the second time, I got the following output:
>   ```
>   > @itwin/property-grid-react@1.3.0 test:e2e
>   > node ../../../common/scripts/run-e2e-with-docker.js property-grid
>   
>   
>   #0 building with "default" instance using docker driver
>   
>   #1 [internal] load build definition from Dockerfile
>   #1 transferring dockerfile: 1.67kB done
>   #1 DONE 0.0s
>   
>   #2 [internal] load .dockerignore
>   #2 transferring context: 76B done
>   #2 DONE 0.0s
>   
>   #3 [internal] load metadata for docker.io/library/debian:11
>   #3 DONE 3.7s
>   
>   #4 [ 1/11] FROM docker.io/library/debian:11@sha256:f33900927c0a8bcf3f0e2281fd0237f4780cc6bc59729bb3a10e75b0703c5ca7
>   #4 DONE 0.0s
>   
>   #5 [internal] load build context
>   #5 transferring context: 702.58kB 0.5s done
>   #5 DONE 0.5s
>   
>   #6 [ 8/11] RUN rush install
>   #6 CACHED
>   
>   #7 [ 2/11] RUN apt-get update   && apt-get install -y curl   && apt-get install -y libnss3 libatk-bridge2.0-0 libasound2   libcairo2 libcups2 libdbus-1-3 libatk1.0-0 libdrm2 libxrandr2   libglib2.0-0 libxkbcommon0 libnspr4 libatspi2.0-0 libgbm1  libpango-1.0-0 libxcomposite1  libxdamage1  libxfixes3  && curl -sL https://deb.nodesource.com/setup_18.x | bash -   && apt-get install -y git nodejs locales zsh procps
>   #7 CACHED
>   
>   #8 [ 3/11] WORKDIR /workspaces/viewer-components-react/
>   #8 CACHED
>   
>   #9 [ 4/11] COPY . .
>   #9 CACHED
>   
>   #10 [ 6/11] RUN git config --local user.name imodeljs-admin
>   #10 CACHED
>   
>   #11 [ 5/11] RUN git config --local user.email imodeljs-admin@users.noreply.github.com
>   #11 CACHED
>   
>   #12 [ 7/11] RUN npm install -g @microsoft/rush
>   #12 CACHED
>   
>   #13 [10/11] WORKDIR /workspaces/viewer-components-react/packages/itwin/property-grid
>   #13 CACHED
>   
>   #14 [ 9/11] RUN rush build
>   #14 CACHED
>   
>   #15 [11/11] RUN npx playwright install
>   #15 CACHED
>   
>   #16 exporting to image
>   #16 exporting layers done
>   #16 writing image sha256:941f74ddb626929538c7482e18d1031c0f11dd1693a26e8365b70a7a9c31a888 done
>   #16 naming to docker.io/library/property-grid-e2e-test-image done
>   #16 DONE 0.0s
>   
>   What's Next?
>     View a summary of image vulnerabilities and recommendations → docker scout quickview
>   
>   Error executing command: Error: Command failed: docker run --name property-grid-e2e-test-container -e UPDATE_SNAPSHOTS=1 property-grid-e2e-test-image
>   docker: Error response from daemon: Conflict. The container name "/property-grid-e2e-test-container" is already in use by container "b3f3e716c36f6b0314f3c616c75f7a9a42b79f9960c1c0eef9e67200f6dde769". You have to remove (or rename) that container to be able to reuse that name.
>   See 'docker run --help'.
>   
>   
>   > @itwin/property-grid-react@1.3.0 test:e2e:local
>   > playwright test
>   
>   
>   
>   
>   Running 4 tests using 4 workers
>     4 passed (44.8s)                                                                                                                                                                                                                                                                                                                               
>   
>   To open last HTML report run:
>   
>     npx playwright show-report
>   
>   [WebServer] (node:127) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.                                                                                                                                                           
>   [WebServer] (node:127) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.            
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   It looks like it succeeded but what's up with that error?
>   ```
>   Error executing command: Error: Command failed: docker run --name property-grid-e2e-test-container -e UPDATE_SNAPSHOTS=1 property-grid-e2e-test-image
>   docker: Error response from daemon: Conflict. The container name "/property-grid-e2e-test-container" is already in use by container "b3f3e716c36f6b0314f3c616c75f7a9a42b79f9960c1c0eef9e67200f6dde769". You have to remove (or rename) that container to be able to reuse that name.
>   See 'docker run --help'.
>   ```
- "Could we forward container's log to host's console?" - this was not happening because "exec" outputs only after the command has finished. Changed to use "spawn" instead of "exec", as it can output during execution.
- " It looks like it succeeded but what's up with that error?" - this is what happened: docker run failed, because you have created a container with the exact same name. After it showed the error message, it logged the output from container (which you have previously created), and the output was success. Added a "finally" block which removes the container which was created.

grigasp commented on common/scripts/run-e2e-with-docker.js:
```diff
@@ -3,49 +3,50 @@
+    // Run Docker container
+    await execute("docker", ["run", "--name", `${dockerContainerName}`, "-e", `UPDATE_SNAPSHOTS=${process.env.UPDATE_SNAPSHOTS}`, `${dockerImageName}`]);
+
+    if(process.env.UPDATE_SNAPSHOTS) {
```
```suggestion
    if (process.env.UPDATE_SNAPSHOTS) {
```

grigasp commented on common/scripts/run-e2e-with-docker.js:
```diff
@@ -3,49 +3,50 @@
\ No newline at end of file
+      await execute("docker", ["cp", `${dockerContainerName}:/workspaces/viewer-components-react/${srcFolderLocation}/e2e-tests`, `./${srcFolderLocation}`]);
+    }
+  } catch(_e) {
```
```suggestion
  } catch {
```

grigasp commented on common/scripts/run-e2e-with-docker.js:
```diff
@@ -3,49 +3,50 @@
+  } catch(_e) {
+    process.exitCode = 1;
+  } finally {
+    // Remove the Docker container
+      await execute("docker", ["rm", "-f", `${dockerContainerName}`]);
```
```suggestion
    // Remove the Docker container
    await execute("docker", ["rm", "-f", `${dockerContainerName}`]);
```

grigasp reviewed (approved):
@aruniverse please have a look

aruniverse commented on common/changes/@itwin/property-grid-react/JonasD-docker-e2e_2023-09-18-16-50.json:
```diff
@@ -0,0 +1,10 @@
+  "changes": [
+    {
+      "packageName": "@itwin/property-grid-react",
+      "comment": "running e2e tests with docker",
```
```suggestion
      "comment": "",
```

aruniverse commented on common/changes/@itwin/tree-widget-react/JonasD-docker-e2e_2023-09-18-16-50.json:
```diff
@@ -0,0 +1,10 @@
+  "changes": [
+    {
+      "packageName": "@itwin/tree-widget-react",
+      "comment": "running e2e tests with docker",
```
```suggestion
      "comment": "",
```
